[dynamo] guarded config

jon-chuang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113929
* #113916
* #113901
* #111300
* __->__ #111299

---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @ezyang @Chillee

jon-chuang committed:
[dynamo] guarded config

[ghstack-poisoned]

jon-chuang commented:
@pytorchbot label "release notes: dynamo"

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: https://github.com/pytorch/pytorch/pull/111145, based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: https://github.com/pytorch/pytorch/pull/111145, based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: https://github.com/pytorch/pytorch/pull/111145, based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: https://github.com/pytorch/pytorch/pull/111145, based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: https://github.com/pytorch/pytorch/pull/111145, based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: https://github.com/pytorch/pytorch/pull/111145, based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: https://github.com/pytorch/pytorch/pull/111145, based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang commented on torch/_dynamo/convert_frame.py:
```diff
@@ -412,6 +415,40 @@ def _clone_with_backend(backend):
     return wrap_convert_context(_convert_frame_assert)
 
 
+def _patch_config_if_changed():
```
Add docbloc? Remove? i.e. always patch - but may be expensive...

jon-chuang replied:
Ok, you know what? Let's get rid of this. It's "premature optimization"

jon-chuang replied:
Actually, this function is not too bad, mostly we are just logging.

jon-chuang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -262,19 +262,61 @@ def innermost_fn(fn):
+def _maybe_init_guarded_config_cache():
+    if not hasattr(config_cache, "saved_config_and_hash"):
+        config_cache.saved_config_and_hash: Optional[  # type: ignore[misc]
+            Tuple[Dict[str, Any], bytes]
```
this is now `Tuple[Dict[str, Any], str]`

jon-chuang replied:
I think NamedTuple would be good, it should be performant too. And it will make code more readable and more difficult to screw up.

jon-chuang replied:
Actually, NamedTuple is very slow when using field accessors. Better to use an object. https://medium.com/@jacktator/dataclass-vs-namedtuple-vs-object-for-performance-optimization-in-python-691e234253b9

jon-chuang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -262,19 +262,61 @@ def innermost_fn(fn):
+            config_cache.saved_config_and_hash = saved_config, saved_config_hash
+            log.debug("Setting top-level compile config hash: %s", saved_config_hash)
+        else:
+            log.debug("Ignoring inner dynamo compile config and hash")  # TODO: remove?
```
this should no longer be verbose, so we don't need to remove/put behind verbose flag

jon-chuang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -516,13 +580,17 @@ def catch_errors_wrapper(callback, hooks: Hooks):
             or config.disable
         ):
-            log.debug("skipping %s %s", frame.f_code.co_name, frame.f_code.co_filename)
+            if not is_skipfile or config.verbose:  # don't log verbose skipfile skips
```
@ezyang continuing from prev discussion this is verbose for the following:

Obtained by running `python -m pytest test/dynamo/test_logging.py -vv -s -k dynamo_debug`

But I think it's because it i

```
test/dynamo/test_logging.py::LoggingTests::test_dynamo_debug <- torch/testing/_internal/logging_utils.py [2023-10-14 16:42:43,872] torch._dynamo.eval_frame: [DEBUG] Saving dynamo config and hash for new compiled object(s). Hash: 9e9ed848cd0751349ffd46d2042c96c5
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping helper /usr/lib/python3.10/contextlib.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping __init__ /usr/lib/python3.10/contextlib.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping __enter__ /usr/lib/python3.10/contextlib.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping backend_cache_wrapper /home/jonch/Desktop/Programming/mlsys/pytorch/torch/_dynamo/eval_frame.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping _maybe_init_guarded_backend_cache /home/jonch/Desktop/Programming/mlsys/pytorch/torch/_dynamo/eval_frame.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping innermost_fn /home/jonch/Desktop/Programming/mlsys/pytorch/torch/_dynamo/eval_frame.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping _set_current_backend /home/jonch/Desktop/Programming/mlsys/pytorch/torch/_dynamo/eval_frame.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping __init__ /usr/lib/python3.10/contextlib.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping __enter__ /usr/lib/python3.10/contextlib.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping restore_guarded_dynamo_config /home/jonch/Desktop/Programming/mlsys/pytorch/torch/_dynamo/eval_frame.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping _maybe_init_guarded_config_cache /home/jonch/Desktop/Programming/mlsys/pytorch/torch/_dynamo/eval_frame.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping debug /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping isEnabledFor /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping _log /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping findCaller /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping <lambda> /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping normcase /usr/lib/python3.10/posixpath.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping makeRecord /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping __init__ /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping __instancecheck__ /usr/lib/python3.10/abc.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping getLevelName /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping basename /usr/lib/python3.10/posixpath.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping _get_sep /usr/lib/python3.10/posixpath.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] skipping splitext /usr/lib/python3.10/posixpath.py
[2023-10-14 16:42:43,873] torch._dynamo.convert_frame: [DEBUG] skipping because no torch.* _splitext             /usr/lib/python3.10/genericpath.py 121
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping current_thread /usr/lib/python3.10/threading.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping name /usr/lib/python3.10/threading.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping current_process /usr/lib/python3.10/multiprocessing/process.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping name /usr/lib/python3.10/multiprocessing/process.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping handle /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping filter /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping callHandlers /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping handle /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping acquire /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping new_emit /home/jonch/Desktop/Programming/mlsys/pytorch/torch/testing/_internal/logging_utils.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping emit /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping format /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping format /home/jonch/Desktop/Programming/mlsys/pytorch/torch/_logging/_internal.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping getLogger /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping getLogger /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping _acquireLock /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping _releaseLock /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping getMessage /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping formatTime /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping is_available /home/jonch/Desktop/Programming/mlsys/pytorch/torch/distributed/__init__.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping is_initialized /home/jonch/Desktop/Programming/mlsys/pytorch/torch/distributed/distributed_c10d.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping WORLD /home/jonch/Desktop/Programming/mlsys/pytorch/torch/distributed/distributed_c10d.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping default_pg /home/jonch/Desktop/Programming/mlsys/pytorch/torch/distributed/distributed_c10d.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping current_trace_id /home/jonch/Desktop/Programming/mlsys/pytorch/torch/_guards.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping get /home/jonch/Desktop/Programming/mlsys/pytorch/torch/_guards.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping <genexpr> /home/jonch/Desktop/Programming/mlsys/pytorch/torch/_logging/_internal.py
[2023-10-14 16:42:43,873] torch._dynamo.eval_frame: [DEBUG] Setting top-level compile config hash: 9e9ed848cd0751349ffd46d2042c96c5
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping flush /usr/lib/python3.10/logging/__init__.py
[2023-10-14 16:42:43,874] torch._dynamo.eval_frame: [DEBUG] skipping release /usr/lib/python3.10/logging/__init__.py
```

Which I do believe is too much when running on this simple function:
```
def example_fn(a):
    output = a.mul(torch.ones(1000, 1000))
    output = output.add(torch.ones(1000, 1000))
    return output
```

The reason seems to be tracing through nested `_TorchDynamoContext` context setup.

jon-chuang replied:
I believe this would even trigger if we trace through `_DisableContext` etc.

jon-chuang replied:
Following up to https://github.com/pytorch/pytorch/pull/111074#discussion_r1355976394

jon-chuang replied:
@ezyang still need your advice on whether we should get rid of this verbose skipfiles logging.

jon-chuang replied:
I'm just going to stack this to defer the changes, for now we will accept the verbosity.

jon-chuang commented on torch/_dynamo/convert_frame.py:
```diff
@@ -215,13 +215,14 @@ def has_tensor(obj):
-        frame.f_code.co_filename,
-        frame.f_code.co_firstlineno,
-    )
+    if config.verbose:
```
maybe this one is fine, it is not as verbose as when we trace through TorchDynamoContext

jon-chuang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -612,6 +682,7 @@ def optimize(
     guard_fail_fn=None,
     disable=False,
     dynamic=None,
+    save_config=True,
```
Please add documentation on this API. See: https://github.com/pytorch/pytorch/pull/111074#issuecomment-1759512216

jon-chuang replied:
Added

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -300,8 +345,19 @@ def __init__(
         self.export = export
```
Please investigate and report whether export can be removed (follow up PR)

jon-chuang replied:
Followed up - seems safe, stacked

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -591,6 +595,15 @@ def BACKEND_MATCH(self, guard: Guard):
+
+        assert guard.source is GuardSource.GLOBAL
+        code = [
+            f"___applicable_config_hash() == '{torch._dynamo.eval_frame.get_saved_else_current_config_hash().hex()}'"
```
why are you hex'ing it 🤔

jon-chuang replied:
Well the choice is between storing the hex value and storing the bytes. I went with storing the bytes but making the guard more human-readable with the hex :sweat_smile:

ezyang replied:
If you repr the bytes string it will still be human readable, but you still avoid unnecessary comparing a lot of extra bytes you don't need

jon-chuang replied:
If this is what you mean by human readable, I unfortunately wouldn't agree? 

```
b'\x92\xec>\xbc\x03\x03\x8c Ol\xd3\xa7\xa2M]0'
```

ezyang replied:
It's ascii, and it uniquely determines the hash, what more can you be looking for :P

ezyang commented:
This is waiting on more improvements from the old PR review request

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

ezyang commented:
Any updates here?

jon-chuang commented:
@ezyang will be cleaning up outstanding PRs over next few days, this one is one of the next.

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang commented on torch/utils/_config_module.py:
```diff
@@ -172,6 +172,21 @@ def codegen_config(self) -> str:
+        """Hashes the configs that are not compile_ignored, along with updates"""
+        if any(k in self._compile_ignored_keys for k in updates):
+            raise ValueError("update keys cannot be @compile_ignored")
+        cfg = dict(self._config)
```
Hmm, I think you need to update this - only compile_ignored should be included

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang commented:
@ezyang apologies for the delay. Should be ready now except for skipfiles logging verbosity confirmation

ezyang commented on test/dynamo/test_config.py:
```diff
@@ -110,6 +110,206 @@ def test_config_hash(self):
+        ):
+            for i in range(2, 12):
+                # Only 4-11 will now be recompiled under old config
+                # 2-3 have been recompiled under old config due to shape mismatch
```
I don't understand this comment, but the test case looks right. Shouldn't 2-3 cache hit from the previous run? The important thing is we DO recompile even though dynamic was requested, because we're using the old config here.

jon-chuang replied:
Yes, I think it means 2-3 have already been (re)compiled. Let just say "should hit cache"

ezyang commented on test/dynamo/test_config.py:
```diff
@@ -110,6 +110,206 @@ def test_config_hash(self):
+        with torch._dynamo.config.patch(
+            automatic_dynamic_shapes=True, assume_static_by_default=False
+        ):
+            opt_fn_static_shape = torch._dynamo.optimize(cnt_dynamic)(fn)
```
misnamed, this is a dynamic shape opt fn

ezyang reviewed (approved):
woohoo!

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym ezyang Chillee


[ghstack-poisoned]

jon-chuang commented:
@pytorchbot merge

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng ezyang Chillee


[ghstack-poisoned]

jon-chuang commented:
@pytorchbot merge

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng ezyang Chillee


[ghstack-poisoned]

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng ezyang Chillee


[ghstack-poisoned]

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng ezyang Chillee


[ghstack-poisoned]

jon-chuang commented:
@pytorchbot merge

jon-chuang committed:
Update on "[dynamo] guarded config"


---

Fixes: https://github.com/pytorch/pytorch/issues/110682

Replaces: https://github.com/pytorch/pytorch/pull/111074

The guards are installed based on config that is valid at the call to `torch.compile`, rather than at any subsequent call / triggered compilation. Subsequent compilations will restore the config if there is a config mismatch of the existing global config with the saved config.

TODO:
- [X] add tests

Follow up PRs:
- [x] add revised cache size computation (follow up PR: #111300 , based on: https://github.com/pytorch/pytorch/pull/107496) 
- [ ] handle run-only mode?
- [ ] config restoration itself is not thread-safe (tracked: https://github.com/pytorch/pytorch/issues/111150)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng ezyang Chillee


[ghstack-poisoned]

jon-chuang commented:
@pytorchbot merge

ezyang commented on patch:
```diff
@@ -0,0 +1,123 @@
+diff --git a/third_party/cutlass b/third_party/cutlass
```
hol up, what is this?!

jon-chuang replied:
I think it's gone now? I accidentally committed a bunch of stray files

davidberard98 commented:
@jon-chuang heads up, we're planning to revert this diff (and a few others that are required for a clean revert).

It causes a significant compile time regression on an internal mode. Still investigating why; I'm hoping to report back once I understand the issue better.

jon-chuang commented:
I think this might be expected. The users may need to be educated on how to properly apply compile time config to their compiled artifacts, perhaps the key is communicating this better.

cc @ezyang @voznesenskym @Chillee as original instigators for this work.

ezyang commented:
Have there been any updates on relanding this?

davidberard98 commented:
I'm still trying to root cause the issue, will comment on the issue
